flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/serialization/SpanningRecordSerializationTest.java

```diff
@@ -105,11 +119,235 @@ public void testHandleMixedLargeRecords() throws Exception {
+			Matcher<? extends Throwable> expectedCause) throws Exception {
+		expectedException.expectCause(expectedCause);
+		testHandleWrongDeserialization(testValue, segmentSize);
+	}
+
+	private void testHandleWrongDeserialization(
+			WrongDeserializationValue testValue,
+			IntFunction<Integer> segmentSizeProvider) throws Exception {
+		int serializedBytes = getSerializedBytes(testValue);
+		int segmentSize = segmentSizeProvider.apply(serializedBytes);
+		testHandleWrongDeserialization(testValue, segmentSize);
+	}
+
+	/**
+	 * Executes the de/serialization round trip test with a serializer that consumes more or less
+	 * bytes than what it writes and verifies the expected exception is thrown.
+	 */
+	private void testHandleWrongDeserialization(
+			WrongDeserializationValue testValue,
+			int segmentSize) throws Exception {
+		List<IOReadableWritable> originalRecords = new ArrayList<>(1);
+		originalRecords.add(testValue);
+
+		expectedException.expect(IOException.class);
+		if (testValue instanceof DeserializingTooMuch) {
+			expectedException.expectMessage(" -1 remaining unread byte");
+		} else if (testValue instanceof DeserializingNotEnough) {
+			expectedException.expectMessage(" 1 remaining unread byte");
+		} else {
+			fail("Invalid test value: " + testValue);
+		}
+
+		testSerializationRoundTrip(originalRecords, segmentSize);
+	}
+
+	/**
+	 * Retrieves the number of bytes the serialized representation of the given value takes (by
+	 * doing the serialization).
+	 */
+	private int getSerializedBytes(IOReadableWritable testValue) throws IOException {
+		SpanningRecordSerializer<IOReadableWritable> serializer = new SpanningRecordSerializer<>();
+		serializer.serializeRecord(testValue);
+		BufferBuilder bufferBuilder = createBufferBuilder();
+		assertTrue("Incorrect test setup: buffer not big enough to contain test value",
+			serializer.copyToBufferBuilder(bufferBuilder).isFullRecord());
+		int writtenBytes = bufferBuilder.finish();
+		bufferBuilder.createBufferConsumer().close();
+		return writtenBytes;
+	}
+
+	/**
+	 * Large object that tries to deserialize an additional byte that it has not written during
+	 * serialization.
+	 */
+	public static class LargeObjectTypeDeserializingTooMuch extends LargeObjectType implements DeserializingTooMuch {
+
+		@SuppressWarnings("WeakerAccess")
+		public LargeObjectTypeDeserializingTooMuch() {
+		}
+
+		@SuppressWarnings("WeakerAccess")
+		public LargeObjectTypeDeserializingTooMuch(int len) {
+			super(len);
+		}
+
+		@Override
+		public void read(DataInputView in) throws IOException {
+			super.read(in);
+			in.readUnsignedByte(); // not written by write()
+		}
+
+		@Override
+		public LargeObjectTypeDeserializingTooMuch getRandom(Random rnd) {
+			int length = super.getRandom(rnd).length();
+			return new LargeObjectTypeDeserializingTooMuch(length);
+		}
+
+		static LargeObjectTypeDeserializingTooMuch getRandom() {
+			return (new LargeObjectTypeDeserializingTooMuch()).getRandom(new Random());
```

Comment:
Do not use completely random random, otherwise tests might start failing randomly and failures might be impossible to reproduce. Either use fixed random seed, or it requires bigger changes in the testing framework, like printing seed in the console and providing a way to re-run the same test with fixed seed.

Reply:
Using a fixed seed is not preferable: you'd always test the same record(s) while otherwise ecah test would test something else and you get better test coverage over time. (I don't know why `org.apache.flink.testutils.serialization.types.Util` is using that pattern though)

I'll have a quick look on whether it is possible to print something generic (the seed) in case of unit test failure - the latter part can be done manually when trying to fix an error.

Reply:
alright: seems like this problem is a bigger task and affects even more things that just this PR: even now, `SpanningRecordSerializationTest` uses two approaches: one `private static final Random RANDOM = new Random(42);` and a `Random rnd = new Random();` in `testHandleMixedLargeRecords()`. I am sure that this test is not the only one with that problem.

I'd suggest to handle this separately.

Reply:
I don't want you to fix other places, but don't make it worse/don't repeat the same mistake. Imagine that this test is flaky on travis and it failed there once with unknown seed and now you have to reproduce this flaky test that fails only 1 out of 2^64 times (`long` seed).

Why fixed seed is bad in this case? You are not touching the content in any way, but just relaying on the record length (right?), so fixed seed shouldn't matter.